src/assets/scripts/client/aircraft/FlightManagementSystem/LegModel.js

```diff
@@ -1,538 +1,780 @@
      */
-    skipToWaypointAtIndex(waypointIndex) {
-        this.waypointCollection = _drop(this.waypointCollection, waypointIndex);
+    _determineLegType(airwayOrProcedureName, navigationLibrary) {
+        if (this._routeString.indexOf('.') === INVALID_NUMBER) {
+            return LEG_TYPE.DIRECT;
+        }
+
+        if (navigationLibrary.hasAirway(airwayOrProcedureName)) {
+            return LEG_TYPE.AIRWAY;
+        }
+
+        if (navigationLibrary.hasProcedure(airwayOrProcedureName)) {
+            return LEG_TYPE.PROCEDURE;
+        }
+
+        throw new TypeError(`Expected airway or procedure name, but we can't ' +
+            'determine what kind of leg ${airwayOrProcedureName} is`
+        );
     }
 
     /**
-     * Destroy the current `WaypointModel` and remove it from the
-     * `waypointCollection`.
-     *
-     * This puts the item that previously occupied the `1` index now
-     * at `0` making it the `currentWaypoint`.
+     * Verify that we are not attempting to initialize this `LegModel` with a route
+     * string that should have been two separate `LegModels`, and throw errors if
+     * we ever DO make such a mistake.
      *
      * @for LegModel
-     * @method moveToNextWaypoint
+     * @method _ensureRouteStringIsSingleSegment
+     * @param routeString {string}
      */
-    moveToNextWaypoint() {
-        this.currentWaypoint.destroy();
-        // this is mutable
-        this.waypointCollection.shift();
+    _ensureRouteStringIsSingleSegment(routeString) {
+        if (routeString.indexOf('..') !== INVALID_INDEX) {
+            throw new TypeError(`Expected single fix or single procedure route string, but received '${routeString}'`);
+        }
+
+        if (routeString.split('.').length > 3) {
+            throw new TypeError(`Expected single procedure route string, but received '${routeString}'`);
+        }
     }
 
     /**
-     * Find if a Waypoint exists within `#waypointCollection`
+     * Generate an array of `WaypointModel`s an aircraft's FMS will need in order to
+     * navigate along this leg instance.
      *
      * @for LegModel
-     * @method hasWaypoint
-     * @param waypointName {string}
-     * @return {boolean}
+     * @param _generateWaypointCollection
+     * @param entryOrFixName {string} name of the entry point (if airway/procedure), or fix name
+     * @param exit {string} name of exit point (if airway/procedure), or `undefined`
+     * @return {array<WaypointModel>}
      */
-    hasWaypoint(waypointName) {
-        // using a for loop here instead of `_find()` because this operation could happen a lot. a for
-        // loop is going to be faster than `_find()` in most cases.
-        for (let i = 0; i < this.waypointCollection.length; i++) {
-            const waypoint = this.waypointCollection[i];
+    _generateWaypointCollection(entryOrFixName, exit) {
+        if (this._legType === LEG_TYPE.DIRECT) {
+            return [new WaypointModel(entryOrFixName)];
```

Comment:
I know why we are returning an array like this, we should add a comment here so anybody else would know too.